### ISSUE TITLE
model.fit() never calls custom metric

### ISSUE THREAD

**Issue by schissmantics** [tensorflow/tensorflow#53009]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): custom code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu / OSX
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.6.0-0-g919f693420e 2.6.0, v2.7.0-rc1-69-gc256c071bb2 2.7.0
- Python version: 3.7, 3.9

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Following the [documentation](https://keras.io/api/metrics/) I'm trying to implement a custom metric, however the metric never gets called in the first place. I added sanity checks that should produce errors when the metric is called. You can find the full example in this [notebook][1]

    def my_metric_fn(y_true, y_pred):
        1 / 0
        while True:
            pass
        squared_difference = tf.square(y_true - y_pred)
        return tf.reduce_mean(squared_difference, axis=-1) 

The code runs without a problem after passing the metric

    model.compile(optimizer=opt, metrics=[my_metric_fn])
    history = model.fit(
    train_dataset,
    validation_data=validation_dataset,
    epochs=epochs,
    callbacks=[early_stopping]
    )

What I actually get:

    Epoch 1/100
    59/59 [==============================] - 27s 451ms/step - loss: 16.3928 - my_metric_fn: 0.0000e+00 - val_loss: 16.5252 - val_my_metric_fn: 0.0000e+00
    Epoch 2/100
    59/59 [==============================] - 25s 420ms/step - loss: 16.3508 - my_metric_fn: 0.0000e+00 - val_loss: 16.5316 - val_my_metric_fn: 0.0000e+00
    Epoch 3/100
    59/59 [==============================] - 25s 420ms/step - loss: 16.3420 - my_metric_fn: 0.0000e+00 - val_loss: 16.5372 - val_my_metric_fn: 0.0000e+00
    Epoch 4/100
    59/59 [==============================] - 25s 417ms/step - loss: 16.3365 - my_metric_fn: 0.0000e+00 - val_loss: 16.5287 - val_my_metric_fn: 0.0000e+00
    Epoch 5/100
    59/59 [==============================] - 25s 418ms/step - loss: 16.3251 - my_metric_fn: 0.0000e+00 - val_loss: 16.5271 - val_my_metric_fn: 0.0000e+00

  [1]: https://colab.research.google.com/drive/1YohoIRqUWuvAQvEot5cFXWRfAPxm82cm?usp=sharing

**Describe the expected behavior**

The metric should be called followed by an error.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**

I included a notebook earlier

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by sanatmpa1** [comment]
@schissmantics,

As you're trying to implement a custom metric which belongs to `Keras`. Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues),To know more see;[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
